packages/core/src/interpreter.ts

```diff
@@ -171,11 +171,22 @@ export class Interpreter<
   }
 
   public get initialState(): State<TContext, TEvent, TStateSchema, TTypestate> {
-    return (
-      this._initialState ||
-      ((this._initialState = this.machine.getInitialState(this.ref)),
-      this._initialState)
-    );
+    try {
+      CapturedState.current = {
+        actorRef: this.ref,
+        spawns: []
+      };
+      return (
+        this._initialState ||
+        ((this._initialState = this.machine.getInitialState()),
+        this._initialState)
+      );
+    } finally {
+      CapturedState.current = {
+        actorRef: undefined,
+        spawns: []
+      };
+    }
```

Comment:
Should we (re)introduce a `withCurrentRef(this.ref, c => ... )` helper function to DRY this up?

Reply:
Maybe, I've mainly wanted to avoid a closure here (which probably ultimately doesn't matter) - gonna also look later if it's possible to somehow drop this when reading the initial state, or just remodel how the initial state is instantiated. I'm not yet sure how to do that and which path to explore but there are some subtle things in terms of how machine API and the interpreter interact with each other - need to start writing things down to open a discussion about this later.

Reply:
Yeah, initial thoughts:

- If interpreter started without context (`service.start()`), then the "cached" initial state can be used
- If interpreter started with context/different state (`service.start(someState)`), then the initial state needs to be reconstructed